sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/OnTimerInvokers.java

```diff
@@ -0,0 +1,270 @@
+            //     this.delegate.<@OnTimer method>(... pass the right args ...)
+            //   }
+            .method(ElementMatchers.named("invokeOnTimer"))
+            .intercept(new InvokeOnTimerDelegation(signature.onTimerMethods().get(timerId)));
+
+    DynamicType.Unloaded<?> unloaded = builder.make();
+
+    @SuppressWarnings("unchecked")
+    Class<? extends OnTimerInvoker<?, ?>> res =
+        (Class<? extends OnTimerInvoker<?, ?>>)
+            unloaded
+                .load(
+                    OnTimerInvokers.class.getClassLoader(),
+                    ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+    return res;
+  }
+
+  /**
+   * An "invokeOnTimer" method implementation akin to @ProcessElement, but simpler because no
+   * splitting-related parameters need to be handled.
+   */
+  private static class InvokeOnTimerDelegation extends DoFnMethodDelegation {
+
+    private final DoFnSignature.OnTimerMethod signature;
+
+    public InvokeOnTimerDelegation(DoFnSignature.OnTimerMethod signature) {
+      super(signature.targetMethod());
+      this.signature = signature;
+    }
+
+    @Override
+    protected StackManipulation beforeDelegation(MethodDescription instrumentedMethod) {
+      // Parameters of the wrapper invoker method:
+      //   ExtraContextFactory.
+      // Parameters of the wrapped DoFn method:
+      //   [OnTimerContext, BoundedWindow, InputProvider, OutputReceiver] in any order
+      ArrayList<StackManipulation> parameters = new ArrayList<>();
+      // Push the extra arguments in their actual order.
+      StackManipulation pushExtraContextFactory = MethodVariableAccess.REFERENCE.loadOffset(1);
+      for (DoFnSignature.Parameter param : signature.extraParameters()) {
+        parameters.add(
+            new StackManipulation.Compound(
+                pushExtraContextFactory,
+                MethodInvocation.invoke(DoFnInvokers.getExtraContextFactoryMethod(param))));
+      }
+      return new StackManipulation.Compound(parameters);
+    }
+  }
+
+  /**
+   * A constructor {@link Implementation} for a {@link DoFnInvoker class}. Produces the byte code
+   * for a constructor that takes a single argument and assigns it to the delegate field.
+   */
+  private static final class InvokerConstructor implements Implementation {
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
+      return instrumentedType;
+    }
+
+    @Override
+    public ByteCodeAppender appender(final Target implementationTarget) {
+      return new ByteCodeAppender() {
+        @Override
+        public Size apply(
+            MethodVisitor methodVisitor,
+            Context implementationContext,
+            MethodDescription instrumentedMethod) {
+          StackManipulation.Size size =
+              new StackManipulation.Compound(
+                      // Load the this reference
+                      MethodVariableAccess.REFERENCE.loadOffset(0),
```

Comment:
I'm not confident that bytebuddy will insert the casts for you when subclassing the first option, since it's not `javac`.
Also, the additional casts will impact performance of the methods, but it might not be a big deal.